### PR TITLE
Faster prompt eval w/ early exit after last layer's kv cache write

### PR DESCRIPTION
**Pull request by ochafik**
Prompt evaluation's only side effect is to build the kv cache for all the layers, so we can exit early at the last layer once we've cached its k & v vectors (and also skip the final rmsnorm & logits computation)

Tested: `make runfast` on M2 Max Mac and adjusted `-n` to only test prompt eval using prompts generated w/ [lorem](https://formulae.brew.sh/formula/lorem):

```bash
./run llama2.c.stories110M.bin -n 336 -i "$( lorem -n 128 )"
./run llama2.c.stories110M.bin -n 1997 -i "$( lorem -n 750 )"
```

| -n | master | this PR | speedup wrt/ master | fast tok (https://github.com/karpathy/llama2.c/pull/251) | this PR + fast tok | speedup wrt/ fast tok |
| -- |  -- |  -- |  -- |  -- |  -- |  -- | 
| 336 | 14.8sec | 13.6sec | -8% | 3.5sec | 2.5sec | -28% (?) |
| 1997 | - | - | - | 38.6sec | 32.2sec | -16% |  

Notes:
- Tokenization is very slow in `master` and dominates the execution time. I've added figures after merging the fast tokenizer from https://github.com/karpathy/llama2.c/pull/251 and using it as baseline
- Some numbers are a bit surprising (excessive speed up), I'll test again when I get a chance.

### CODE REVIEW

**Comment by karpathy** [comment]
not willing to bloat things just yet for this speedup

**Comment by ochafik** [comment]
@karpathy thanks for taking a look (and for the cool repo!). Note that after https://github.com/karpathy/llama2.c/pull/251 it would be a 11-16% speedup for 10 extra LOC (updated figures), but fair enough ✌️
